### PR TITLE
automatically associate inputs, labels and messages

### PR DESCRIPTION
**Pull request by mayank99**
## Changes

If a user forgets to connect `Label`/`Input`/`StatusMessage` inside `InputGrid`, we will now handle it for them by cloning `children`.

Since `InputGrid` is also used internally in `LabeledInput`/`LabeledSelect`, I was able to remove some of the explicit id handling in those components.

Lastly, the `StatusMessage` being connected via `aria-describedby` is brand new functionality that we weren't handling before at all, so this is a net improvement for all uses of the `message` prop.

Implementation notes:
- Looping/cloning children is generally [discouraged](https://react.dev/reference/react/Children) and indeed can be quite brittle (breaks if there are wrapper divs anywhere), but it is still a very useful fallback and just a convenient thing.
- It helps ease the v2→v3 migration specifically for the more complex `LabeledInput`→`InputWithDecorations` scenarios (discovered [here](https://github.com/iTwin/iTwinUI-migration-tool/pull/96#discussion_r1446653151)) and especially for React 17 users (who don't have access to `useId`).
- I created an internal `InputWithIcon` component to help with narrowing down children. It's starting to get a little confusing now along with the existing `InputGrid`, `InputContainer`, `InputFlexContainer`, `InputWithDecorations`, `InputGroup` components. We should go through all of them (in a future PR) and try to remove some of them.

## Testing

Made sure existing unit and axe tests pass. Added new unit tests. Also tested a few different permutations of the following code in Vite playground:

```jsx
<InputGrid>
  <Label>Label</Label>
  <Input />
  <StatusMessage>Hint message</StatusMessage>
</InputGrid>
```

Tested the components in screen-readers and the browser's accessibility tree to make sure labels and message are correctly associated and announced.

Works fine in NVDA, but VoiceOver has a bug where it [doesn't automatically announce descriptions on text fields](https://bugs.webkit.org/show_bug.cgi?id=262895#c6).

<details><summary>NVDA 2023.3.1 + Firefox 121</summary>

image

image

</details> 

<details><summary>VoiceOver + Safari 17.0</summary>
<p>

image

image

image

</p>
</details> 

## Docs

Added changesets.

Updated `InputGrid` documentation and jsdocs.

### CODE REVIEW

File: packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx
Comment (rohan-kadkol) [inline packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx:8]: Is it necessary to provide a `displayName` when the name is the same as the component name?

File: packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx
Comment (rohan-kadkol) [inline packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx:7]: Should this be marked as `@private`?

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (rohan-kadkol) [inline packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:39]: Should this mention that if the user passes manual ids, that won't be overriden? Or maybe that can be implied?

Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:39]: It's mentioned in the actual documentation

File: packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx:7]: should this comment be marked as "nit"? 😅

File: packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx:8]: Yes, component name is usually not enough especially when using arrow functions and/or `forwardRef`


React devtools will show "ForwardRefComponent" for the following snippet:
```js
const Stuff = React.forwardRef((props, ref) => {});
```

File: packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx:7]: Since I don't see any changeset saying that `InputWithIcon` is publically exposed, I think this was meant to be a private component. If so, then I think `@private` is important.

Comment (mayank99) [reply packages/itwinui-react/src/core/utils/components/InputWithIcon.tsx:7]: It's not important because it is not exposed in the barrel file. The user cannot import it.

The whole `@private` thing used to only matter in previous major versions when we were exposing internal sub-paths

**Review by rohan-kadkol — COMMENTED** [review]
InputGrid related stories can be updated to remove explicit ids. Examples:
* `ComboBox → With Label`
* `InputGrid` stories
* `Label → Inline`

**Review by mayank99 — COMMENTED** [review]
> InputGrid related stories can be updated to remove explicit ids.

There is no benefit to this, but removed ids from some of them. I kept it in some more complex cases.

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (rohan-kadkol) [inline packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:172]: To be safe, should we stop the `forEach()` after we find the first input?

File: packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx
Comment (rohan-kadkol) [inline packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx:63]: Should we also test cases like:
* When user id is given, we should use that instead
* When `<StatusMessage>` is not provided, there should be no description
* `labelId` should also be there when `hasSelect`
* etc.

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:228]: no, those are just regular form fields so they work with the general input case

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx]: good catch, moved it out of the loop

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:172]: i'm not sure about this; would there ever be more than one input?

there isn't a way to early exit from a `forEach` so i would need to use a `for..of` which requires converting the children to an array (i assume `Children.forEach` has some optimizations that i would be throwing away)

Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx]: actually I realized something:
since i'm no longer using state, it gets reset to `undefined` on every render anyway, so i can just use `||=` to only set it once on every render.

change made: [`0338ecb`](https://github.com/iTwin/iTwinUI/pull/1777/commits/0338ecbcfe395b6d27db731264d9b0cc92a47584)

File: packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx]: >     * When user id is given, we should use that instead

Added tests for this.  [`b8952de`](https://github.com/iTwin/iTwinUI/pull/1777/commits/b8952dee31b8de3a751dbf9063bef54af20b1bbe)

I also changed the code to handle explicitly-passed `undefined`.

```jsx
<InputGrid>
  <Label id={undefined}>Label</Label>
  <Input id={undefined} />
  <StatusMessage id={undefined}>Hint message</StatusMessage>
</InputGrid>
```

>     * When `<StatusMessage>` is not provided, there should be no description

I don't see why this is necessary. We don't care that there is no description.


>     * `labelId` should also be there when `hasSelect` 
>     * etc.

These are implementation details that should not be tested. We only care about the final result, i.e. accName and accDesc are present.

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx]: > i'm not sure about this; would there ever be more than one input?

I guess the officially supported case is with only one input, as demonstrated on our "Input with Decorations" docs page. So if more than one input is included for some reason, it's good that we're only considering the first one (after [`0338ecb` (#1777)](https://github.com/iTwin/iTwinUI/pull/1777/commits/0338ecbcfe395b6d27db731264d9b0cc92a47584) & https://github.com/iTwin/iTwinUI/pull/1777/commits/ed62b3e1ee8d4f98316c5f59ab2bf10bc479db3b)

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:164]: what does this have to do with Select?

Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:164]: that's also unrelated, you might be confusing two different parts:

- L185: We only want `labelId` for `Select`, because it gets referenced in `aria-labelledby` (other inputs use `<Label htmlFor`>)
- L150: We exclude `Select` because its `id` gets set on the wrapping `<div>`. To get its real id, we would have to grab it from `triggerProps`, similar to how we do for ComboBox (L149), but it doesn't matter because we don't even return `inputId` for Select (L186).

File: packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx]: > > `labelId` should also be there when `hasSelect`
> > etc.
> 
> These are implementation details that should not be tested. We only care about the final result, i.e. accName and accDesc are present.

I mean a test to confirm that id is not added to `<Label>` unnecessarily. Instead, it is only added when there is a `<Select>`.

Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx]: It still doesn't matter, it's an implementation detail. We do not care if the Label has an id, and the developer using this component certainly does not care either.

I recommend reading this article: https://kentcdodds.com/blog/testing-implementation-details

File: packages/itwinui-react/src/core/InputGrid/InputGrid.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.tsx:168]: added some comments [`ab5c823`](https://github.com/iTwin/iTwinUI/pull/1777/commits/ab5c823f3e3e477b3eef6e0ad6618cf827c6f607)

File: packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx]: I read that article, but I am not suggesting testing implementation details, like the value of `hasSelect`. I am suggesting having a test to make sure we don't add props when they are not needed.

Even though developers may not care whether an id is added or not, I feel adding it unnecessarily is still incorrect, and thus should be tested that it is not happening.

Basically, since you have written logic to prevent this `id` from being added to `Label` when no `Select` is present, I feel this case/logic should be tested (independent of the implementation details needed to implement this case/logic).

Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/InputGrid/InputGrid.test.tsx]: As discussed via chat, we can skip this test for the reasons you mentioned in this thread 👍

**Review by rohan-kadkol — APPROVED** [review]
LGTM 🎉

I am about to complete the remaining comments from #1753 (base branch for this PR). So, I think it would be better to merge your PR after mine.

If you feel the same, you might prefer to put your PR back to draft after you get the needed PR approvals to prevent accidentally merging your PR before mine.

**Comment by rohan-kadkol** [comment]
> Some unit test cases are failing.

That would likely be from my base PR #1753. After I fix it in my PR, this will likely be fixed in this PR too.

**Comment by rohan-kadkol** [comment]
I recently merged the PR of the base branch of this PR. So, this PR will now be unblocked 👍

**Pull request merged**
